Fix MidiHeader structure.

DKingAlpha opened the pull request:
SDK 10.0.16299.0  mmeapi.h  Ln.1029
```
/* MIDI data block header */
typedef struct midihdr_tag {
    LPSTR       lpData;               /* pointer to locked data block */
    DWORD       dwBufferLength;       /* length of data in data block */
    DWORD       dwBytesRecorded;      /* used for input only */
    DWORD_PTR   dwUser;               /* for client's use */
    DWORD       dwFlags;              /* assorted flags (see defines) */
    struct midihdr_tag far *lpNext;   /* reserved for driver */
    DWORD_PTR   reserved;             /* reserved for driver */

#if (WINVER >= 0x0400)
    DWORD       dwOffset;             /* Callback offset into buffer */
    DWORD_PTR   dwReserved[8];        /* Reserved for MMSYSTEM */
#endif
} MIDIHDR, *PMIDIHDR, NEAR *NPMIDIHDR, FAR *LPMIDIHDR;
```

Commit:
Fix reservedArray counts to 8

SDK 10.0.16299.0  mmeapi.h  Ln.1029
/* MIDI data block header */
typedef struct midihdr_tag {
    LPSTR       lpData;               /* pointer to locked data block */
    DWORD       dwBufferLength;       /* length of data in data block */
    DWORD       dwBytesRecorded;      /* used for input only */
    DWORD_PTR   dwUser;               /* for client's use */
    DWORD       dwFlags;              /* assorted flags (see defines) */
    struct midihdr_tag far *lpNext;   /* reserved for driver */
    DWORD_PTR   reserved;             /* reserved for driver */

#if (WINVER >= 0x0400)
    DWORD       dwOffset;             /* Callback offset into buffer */
    DWORD_PTR   dwReserved[8];        /* Reserved for MMSYSTEM */
#endif
} MIDIHDR, *PMIDIHDR, NEAR *NPMIDIHDR, FAR *LPMIDIHDR;

tebjan commented:
hello, thanks for checking this issues. how does this relate to x86 and x64? is the array still the same size for both?

DKingAlpha commented:
**EDIT:**
I actually ran my win-api version and found that, the size of MIDIHDR **IS** different between x86 and x64.
x86: 0x40
x64: 0x70

Reason:
```
+--------------+------+------+
|              | x86  | x64  |
+--------------+------+------+
| DWORD        | 4    | 4    |
+--------------+------+------+
| DWORD_PTR    | 4    | 8    |
+--------------+------+------+
| LPSTR        | 4    | 8    |
+--------------+------+------+
| midihdr_tag* | 4    | 8    |
+--------------+------+------+
| MIDIHDR      | 0x40 | 0x70 |
+--------------+------+------+
```

Fix: Another PR is on the way.

Original reply:

_Ahh I dont think they are arch related._

_DWORD and LPSTR are all restricted to 32 bits, and I cant find a macro around defines that they only work in x86. So I'd like to believe that they work in both x86 and x64._


[**This**](https://msdn.microsoft.com/en-us/library/6sehtctf.aspx) document lists different windows version.

This issue is found when I try to find the reason why my launchpad wont light up. So I traced down to the win api level and compared it with my previous project(direct winapi, c++, working well), and accidently found the difference.
The reason finally is that I opened wrong output device id. pretty simple mistake.

DKingAlpha commented:
I am try to check the size but found a weird thing.

The interpreted struct should be
```
[StructLayout(LayoutKind.Sequential)]
internal struct MidiHeader
{
    public IntPtr data;
    public Int32 bufferLength;
    public Int32 bytesRecorded;
    public IntPtr user;
    public Int32 flags;
    public IntPtr next;
    public IntPtr reserved;
    public Int32 offset;

    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 8)]
    public IntPtr[] reservedArray;
}
```

In x86, C# works normal, SizeOfMidiHeader=0x40 as expected.

But in x64, SizeOfMidiHeader becomes 0x78=120, which should be 0x70=112.

I believe this is related to DWORD align something. Still testing.

DKingAlpha commented:
```
    internal struct MidiHeader
    {
        //public IntPtr data;
        public Int32 bufferLength;
        public Int32 bytesRecorded;
        //public IntPtr user;
        public Int32 flags;
        //public IntPtr next;
        //public IntPtr reserved;
        public Int32 offset;

        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 8)]
        public IntPtr[] reservedArray;
    }
```
Size 80, correct.

```
    internal struct MidiHeader
    {
        public IntPtr data;           // Uncomment
        public Int32 bufferLength;
        public Int32 bytesRecorded;
        //public IntPtr user;
        public Int32 flags;
        //public IntPtr next;
        //public IntPtr reserved;
        public Int32 offset;

        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 8)]
        public IntPtr[] reservedArray;
    }
```
Size 88, correct.

```
    internal struct MidiHeader
    {
        public IntPtr data;
        public Int32 bufferLength;
        public Int32 bytesRecorded;
        public IntPtr user;           // Uncomment
        public Int32 flags;
        //public IntPtr next;
        //public IntPtr reserved;
        public Int32 offset;

        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 8)]
        public IntPtr[] reservedArray;
    }
```
Size 96, correct.

```
    internal struct MidiHeader
    {
        public IntPtr data;
        public Int32 bufferLength;
        public Int32 bytesRecorded;
        public IntPtr user;
        public Int32 flags;
        public IntPtr next;           // Uncomment
        //public IntPtr reserved;
        public Int32 offset;

        [MarshalAs(UnmanagedType.ByValArray, SizeConst = 8)]
        public IntPtr[] reservedArray;
    }
```
Size 112, incorrect.  Increasing 8B from nowhere.


Additianal test:
Directly uncomment **next** or **reserved**, the problem is reproduced. So these two members are to blame but why...
How are we supposed to do with it?

DKingAlpha commented:
Problem solved by adding  Pack = 1 attibute on structlayout

Commit:
Fix MIDIHDR implement in x64

Some element types are actually not Int32(DWORD 32b) but IntPtr(DWORD_PTR 64b). Fixed.
Tho they may be not affecting because these elements are not used, but anyway, they are standard MIDIHDR now.

tebjan commented:
good find, thanks a lot!

tebjan commented:
unfortunately this seems to crash on startup in a use case i have... the type changes seem not to be correct and the native to managed transition crashed somewhere and i could not find the cause. i'll revert the changes and maybe you can do more testing with midi input and output devices...
